rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -81,7 +81,7 @@ class ChargingReducer[M[_]](implicit R: Reduce[M], C: CostAccountingAlg[M]) {
 
 object Reduce {
 
-  private def substituteAndCharge[A: Chargeable, M[_]: Substitute[?[_], A]: Sync](
+  def substituteAndCharge[A: Chargeable, M[_]: Substitute[?[_], A]: Sync](
```

Comment:
I don't agree to make these public in order to be tested. 

What we need to test here for cost accounting is `costAccountingAlg.charge(Cost(Chargeable[A].cost(term)))` not the whole thing

Reply:
We want to test that substitution works for failure and success. Not that we charge proper cost.

Reply:
still, these are implementation details. The API of the reducer should not be affected by our testing needs. If that is the case then our architecture is not ok

Reply:
I've been there and had this discussion like many times with Artur. This is not refactoring PR.

Reply:
And they aren't implementation details. This is exactly the logic of this method. We have different methods for pure substitution which we test elsewhere.

Reply:
The reducer is doing also cost accounting. To separate that you would need at least a sprint.

Reply:
aftet the discussions we had I understand what you mean that this is not part of the reducer interface but we should then move these functions to the Substitute object. They will be public there and would be tested separately from the reducer. This way everybody is happy

Reply:
basically what I'm saying is that this logic:

```scala
 def substituteAndCharge(...)=
   Substitute[M, A]
      .substitute(term)(depth, env)
      .attempt
      .flatMap(
        _.fold(
          th => // On error charge for the initial term
            (th, Cost(Chargeable[A].cost(term)))
          substTerm =>
            (substTerm, Cost(Chargeable[A].cost(substTerm))) 
        )
      )
```

should go to substitute it also becomes simpler as it's not aware of the M

Reply:
but it's changing the original logic because 
```scala
Substitute[M, A]
      .substitute(term)(depth, env)
```
failed and already did `F.raiseError(…)` and now you take that error and return it as a tuple. 

Also, this implementation will make the usage of it unnecessarily complex and we loose the expressiveness that comes from monadic chain.

Reply:
You are right. I wasn't careful with the types. The charging code is dependent on M. I just copied the piece of code that I thought would be the reusable part. 

See the code below which should be ok:

```scala

  def charge[A : Chargeable, M[_] : Sync](term : A,
                                          action : A => M[A],
                                          costAccountingAlg: CostAccountingAlg[M]) : M[A] =
    action(term)
      .attempt
      .map(
        _.fold(
          th => (Left(th), Cost(Chargeable[A].cost(term))),
          substTerm => (Right(substTerm), Cost(Chargeable[A].cost(substTerm)))))
      .flatMap({case (result, cost) => costAccountingAlg.charge(cost) *> Sync[M].pure(result)})
      .rethrow

  def substituteAndCharge[A: Chargeable, M[_]: Substitute[?[_], A]: Sync](
      term: A,
      depth: Int,
      env: Env[Par],
      costAccountingAlg: CostAccountingAlg[M]
  ): M[A] =
      charge(term,
            Substitute[M, A].substitute(_)(depth, env),
            costAccountingAlg)
```

In this code the function `charge` is the reusable code that can be extracted and unit-tested outside the reducer and then the reducer only adds the choice of choosing the substitute policy on top of it. 

Probably the function `spatialMatchAndCharge` could be written in terms of this `charge` as well.

Reply:
`spatialMatchAndCharge` is bit different, because the notion of failure is totally something else there.
Also, I don't like that you change the error handling logic:
```scala
 .map(
        _.fold(
          th => (Left(th), Cost(Chargeable[A].cost(term))),
          substTerm => (Right(substTerm), Cost(Chargeable[A].cost(substTerm)))))
      .flatMap({case (result, cost) => costAccountingAlg.charge(cost) *> Sync[M].pure(result)})
      .rethrow
```
The original implementation is simpler IMO. 

I don't like  the idea of the rewrite because we have duplicated 2 lines of code. I don't think that this method is tested incorrectly.

Reply:
I'm ok with this